Patch to new recent bindnode PR seems to cause panic in go-graphsync

After updating go-graphsync to go-ipld-prime master, I experience the following panic:

```
=== RUN   TestGraphsyncRoundTripIgnoreCids
panic: interface conversion: basicnode.plainList is not datamodel.Node: missing method Length [recovered]
	panic: interface conversion: basicnode.plainList is not datamodel.Node: missing method Length [recovered]
	panic: interface conversion: basicnode.plainList is not datamodel.Node: missing method Length

goroutine 1865 [running]:
go.opentelemetry.io/otel/sdk/trace.(*recordingSpan).End(0x140003f6900, {0x0, 0x0, 0x0})
	/Users/hannah/projects/go/pkg/mod/go.opentelemetry.io/otel/sdk@v1.2.0/trace/span.go:283 +0x844
panic({0x1054b1b00, 0x140033403c0})
	/opt/homebrew/opt/go/libexec/src/runtime/panic.go:1038 +0x21c
go.opentelemetry.io/otel/sdk/trace.(*recordingSpan).End(0x140003f6a80, {0x0, 0x0, 0x0})
	/Users/hannah/projects/go/pkg/mod/go.opentelemetry.io/otel/sdk@v1.2.0/trace/span.go:283 +0x844
panic({0x1054b1b00, 0x140033403c0})
	/opt/homebrew/opt/go/libexec/src/runtime/panic.go:1038 +0x21c
github.com/ipld/go-ipld-prime/node/bindnode.(*_mapIterator).Next(0x1400334e200)
	/Users/hannah/projects/go/pkg/mod/github.com/ipld/go-ipld-prime@v0.14.5-0.20220306224815-9e783aa21c81/node/bindnode/node.go:1124 +0x3cc
github.com/ipld/go-ipld-prime/node/bindnode.(*_mapIteratorRepr).Next(0x1400334e200)
	/Users/hannah/projects/go/pkg/mod/github.com/ipld/go-ipld-prime@v0.14.5-0.20220306224815-9e783aa21c81/node/bindnode/repr.go:253 +0x28
github.com/ipld/go-ipld-prime/codec/dagcbor.marshalMap({0x1055d1538, 0x14003340300}, 0x14000270230, {0x1055a9240, 0x1400321a280}, {0x1, 0x2})
	/Users/hannah/projects/go/pkg/mod/github.com/ipld/go-ipld-prime@v0.14.5-0.20220306224815-9e783aa21c81/codec/dagcbor/marshal.go:178 +0x380
github.com/ipld/go-ipld-prime/codec/dagcbor.marshal({0x1055d1538, 0x14003340300}, 0x14000270230, {0x1055a9240, 0x1400321a280}, {0x1, 0x2})
	/Users/hannah/projects/go/pkg/mod/github.com/ipld/go-ipld-prime@v0.14.5-0.20220306224815-9e783aa21c81/codec/dagcbor/marshal.go:69 +0x56c
github.com/ipld/go-ipld-prime/codec/dagcbor.marshalMap({0x1055d1538, 0x14003340240}, 0x14000270230, {0x1055a9240, 0x1400321a280}, {0x1, 0x2})
	/Users/hannah/projects/go/pkg/mod/github.com/ipld/go-ipld-prime@v0.14.5-0.20220306224815-9e783aa21c81/codec/dagcbor/marshal.go:214 +0x6cc
github.com/ipld/go-ipld-prime/codec/dagcbor.marshal({0x1055d1538, 0x14003340240}, 0x14000270230, {0x1055a9240, 0x1400321a280}, {0x1, 0x2})
	/Users/hannah/projects/go/pkg/mod/github.com/ipld/go-ipld-prime@v0.14.5-0.20220306224815-9e783aa21c81/codec/dagcbor/marshal.go:69 +0x56c
github.com/ipld/go-ipld-prime/codec/dagcbor.marshal({0x1055d1538, 0x14003340210}, 0x14000270230, {0x1055a9240, 0x1400321a280}, {0x1, 0x2})
	/Users/hannah/projects/go/pkg/mod/github.com/ipld/go-ipld-prime@v0.14.5-0.20220306224815-9e783aa21c81/codec/dagcbor/marshal.go:84 +0x770
github.com/ipld/go-ipld-prime/codec/dagcbor.marshalMap({0x1055d1538, 0x140033401e0}, 0x14000270230, {0x1055a9240, 0x1400321a280}, {0x1, 0x2})
	/Users/hannah/projects/go/pkg/mod/github.com/ipld/go-ipld-prime@v0.14.5-0.20220306224815-9e783aa21c81/codec/dagcbor/marshal.go:214 +0x6cc
github.com/ipld/go-ipld-prime/codec/dagcbor.marshal({0x1055d1538, 0x140033401e0}, 0x14000270230, {0x1055a9240, 0x1400321a280}, {0x1, 0x2})
	/Users/hannah/projects/go/pkg/mod/github.com/ipld/go-ipld-prime@v0.14.5-0.20220306224815-9e783aa21c81/codec/dagcbor/marshal.go:69 +0x56c
github.com/ipld/go-ipld-prime/codec/dagcbor.marshalMap({0x1055d1538, 0x140033401b0}, 0x14000270230, {0x1055a9240, 0x1400321a280}, {0x1, 0x2})
	/Users/hannah/projects/go/pkg/mod/github.com/ipld/go-ipld-prime@v0.14.5-0.20220306224815-9e783aa21c81/codec/dagcbor/marshal.go:214 +0x6cc
github.com/ipld/go-ipld-prime/codec/dagcbor.marshal({0x1055d1538, 0x140033401b0}, 0x14000270230, {0x1055a9240, 0x1400321a280}, {0x1, 0x2})
	/Users/hannah/projects/go/pkg/mod/github.com/ipld/go-ipld-prime@v0.14.5-0.20220306224815-9e783aa21c81/codec/dagcbor/marshal.go:69 +0x56c
github.com/ipld/go-ipld-prime/codec/dagcbor.Marshal(...)
	/Users/hannah/projects/go/pkg/mod/github.com/ipld/go-ipld-prime@v0.14.5-0.20220306224815-9e783aa21c81/codec/dagcbor/marshal.go:57
github.com/ipld/go-ipld-prime/codec/dagcbor.EncodeOptions.Encode({0x1, 0x2}, {0x1055d1538, 0x140033401b0}, {0x1055a85c0, 0x14003340180})
	/Users/hannah/projects/go/pkg/mod/github.com/ipld/go-ipld-prime@v0.14.5-0.20220306224815-9e783aa21c81/codec/dagcbor/marshal.go:45 +0x284
github.com/ipld/go-ipld-prime/codec/dagcbor.Encode(...)
	/Users/hannah/projects/go/pkg/mod/github.com/ipld/go-ipld-prime@v0.14.5-0.20220306224815-9e783aa21c81/codec/dagcbor/multicodec.go:47
github.com/ipfs/go-graphsync/message/v2.(*MessageHandler).ToNet(0x105daba40, {0x14003427870, 0xb}, {0x140033147e0, 0x14003314990, 0x14003314810}, {0x12d218120, 0x14003355940})
	/Users/hannah/projects/go/src/github.com/ipfs/go-graphsync/message/v2/message.go:142 +0x11c
github.com/ipfs/go-graphsync/network.msgToStream({0x1055bbb78, 0x1400334e680}, {0x1055cebc8, 0x14003355940}, 0x14003355c80, {0x140033147e0, 0x14003314990, 0x14003314810}, 0x8bb2c97000)
	/Users/hannah/projects/go/src/github.com/ipfs/go-graphsync/network/libp2p_impl.go:129 +0x454
github.com/ipfs/go-graphsync/network.(*streamMessageSender).SendMsg(0x14003355960, {0x1055bbb78, 0x1400334e680}, {0x140033147e0, 0x14003314990, 0x14003314810})
	/Users/hannah/projects/go/src/github.com/ipfs/go-graphsync/network/libp2p_impl.go:114 +0x6c
github.com/ipfs/go-graphsync/messagequeue.(*MessageQueue).attemptSendAndRecovery(0x14000200300, {0x140033147e0, 0x14003314990, 0x14003314810}, {{0x1055bbc20, 0x140033147b0}, {0x14003314960, 0x140033149c0}, 0x0, 0x0, ...})
	/Users/hannah/projects/go/src/github.com/ipfs/go-graphsync/messagequeue/messagequeue.go:304 +0x64
github.com/ipfs/go-graphsync/messagequeue.(*MessageQueue).sendMessage(0x14000200300)
	/Users/hannah/projects/go/src/github.com/ipfs/go-graphsync/messagequeue/messagequeue.go:252 +0x874
github.com/ipfs/go-graphsync/messagequeue.(*MessageQueue).runQueue(0x14000200300)
	/Users/hannah/projects/go/src/github.com/ipfs/go-graphsync/messagequeue/messagequeue.go:159 +0x70
created by github.com/ipfs/go-graphsync/messagequeue.(*MessageQueue).Startup
	/Users/hannah/projects/go/src/github.com/ipfs/go-graphsync/messagequeue/messagequeue.go:142 +0x3c
FAIL	github.com/ipfs/go-graphsync/impl	0.739s
FAIL
```

it appears the latest code for the line immediately about the panic (https://github.com/ipld/go-ipld-prime/blob/master/node/bindnode/node.go#L1124) was introduced by #371 .

--- a/node/bindnode/infer.go
+++ b/node/bindnode/infer.go
@@ -135,8 +135,17 @@ func verifyCompatibility(seen map[seenEntry]bool, goType reflect.Type, schemaTyp
 		if fieldValues.Type.Kind() != reflect.Map {
 			doPanic("kind mismatch; need struct{Keys []K; Values map[K]V}")
 		}
-		verifyCompatibility(seen, fieldValues.Type.Key(), schemaType.KeyType())
-		verifyCompatibility(seen, fieldValues.Type.Elem(), schemaType.ValueType())
+		keyType := fieldValues.Type.Key()
+		verifyCompatibility(seen, keyType, schemaType.KeyType())
+
+		elemType := fieldValues.Type.Elem()
+		if schemaType.ValueIsNullable() {
+			if elemType.Kind() != reflect.Ptr {
+				doPanic("nullable types must be pointers")
+			}
+			elemType = elemType.Elem()
+		}
+		verifyCompatibility(seen, elemType, schemaType.ValueType())
 	case *schema.TypeStruct:
 		if goType.Kind() != reflect.Struct {
 			doPanic("kind mismatch; need struct")

--- a/node/bindnode/infer_test.go
+++ b/node/bindnode/infer_test.go
@@ -322,6 +322,12 @@ func TestPrototypePointerCombinations(t *testing.T) {
 		{"Link_Generic", "Link", (*datamodel.Link)(nil), `{"/": "bafybeigdyrzt5sfp7udm7hu76uh7y26nf3efuylqabf3oclgtqy55fbzdi"}`},
 		{"List_String", "[String]", (*[]string)(nil), `["foo", "bar"]`},
 		{"Any_Node_Int", "Any", (*datamodel.Node)(nil), `23`},
+		// TODO: reenable once we don't require pointers for nullable
+		// {"Any_Pointer_Int", "{String: nullable Any}",
+		// 	(*struct {
+		// 		Keys   []string
+		// 		Values map[string]datamodel.Node
+		// 	})(nil), `{"x":3,"y":"bar","z":[2.3,4.5]}`},
 		{"Map_String_Int", "{String:Int}", (*struct {
 			Keys   []string
 			Values map[string]int64
@@ -879,6 +885,23 @@ var verifyTests = []struct {
 				Keys   []string
 				Values string
 			})(nil), `.*type Root .*: kind mismatch;.*`},
+		},
+	},
+	{
+		name:      "MapNullableAny",
+		schemaSrc: `type Root {String:nullable Any}`,
+		goodTypes: []interface{}{
+			(*struct {
+				Keys   []string
+				Values map[string]*datamodel.Node
+			})(nil),
+		},
+		badTypes: []verifyBadType{
+			{(*string)(nil), `.*type Root .* type string: kind mismatch;.*`},
+			{(*struct {
+				Keys   []string
+				Values map[string]datamodel.Node
+			})(nil), `.*type Root .*: nullable types must be pointers`},
 		},
 	},
 	{

